subprojects/core-api/src/main/java/org/gradle/api/artifacts/result/ComponentSelectionReason.java

```diff
@@ -83,4 +83,13 @@
      */
     List<ComponentSelectionDescriptor> getDescriptions();
 
+    /**
+     * Informs whether the selected component version has been influenced by a dependency constraint.
+     *
+     * @return true if a dependency constraint influenced the selection of this component
+     *
+     * @since 4.6
+     */
+    @Incubating
+    boolean isConstrained();
```

Comment:
The model is actually pretty simple, but we couldn't rename for binary backwards compatibility. Basically:

- a reason consists of descriptors
- a descriptor consists of a cause and, optionally, a custom message
- a cause if one of the selection causes, like `REQUESTED`, `SELECTED_BY_RULE`, ...

Before the refactor, it was extremely painful to find out what really happened because of that boolean flag set. Now the booleans are just convenience methods.